overlord/snapstate/refreshhints.go

```diff
@@ -0,0 +1,85 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapstate
+
+import (
+	"time"
+
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/store"
+)
+
+var refreshHintsDelay = time.Duration(24 * time.Hour)
+
+// refreshHints will ensure that we regular get data about refreshes
+// so that we can potentially warn the user about importand missing
+// refreshes.
+type refreshHints struct {
+	state *state.State
+}
+
+func newRefreshHints(st *state.State) *refreshHints {
+	return &refreshHints{state: st}
+}
+
+func (r *refreshHints) lastRefresh() (time.Time, error) {
+	var lastRefresh time.Time
+	if err := r.state.Get("last-refresh-hints", &lastRefresh); err != nil && err != state.ErrNoState {
+		return time.Time{}, err
+	}
+	return lastRefresh, nil
+}
+
+func (r *refreshHints) needsUpdate() (bool, error) {
+	t, err := r.lastRefresh()
+	if err != nil {
+		return false, err
+	}
+	return t.Before(time.Now().Add(-refreshHintsDelay)), nil
+}
+
+func (r *refreshHints) refresh() error {
+	refreshManaged := false
+
+	_, _, _, err := refreshCandidates(r.state, nil, nil, &store.RefreshOptions{RefreshManaged: refreshManaged})
+	r.state.Set("last-refresh-hints", time.Now())
```

Comment:
Should we set this only if `err != nil`?

Reply:
Maybe, a good question. catalog-refresh does also not care about the error. If we only set if err == nil then we need to ensure that there is a delay (back-off) between each attempts (as ensure can be called a lot). As this is purely informal I think its ok, but I'm happy to implement it different if there is agreement on it.

Reply:
I think for now this is ok, maybe worth a TODO about retrying a few times with backoff later

Reply:
Good point, I added a TODO